File: b897(find_the_ways)/C++/find_the_ways.cpp
#define _USE_MATH_DEFINES
#include <cmath>
#include <iostream>


/* logorithm of Stirling formula */
inline long double FactLog(const long long n) {
  return log10(sqrt(2 * n * M_PI)) + log10(n / M_E) * n;
}


int main(int argc, char const *argv[]) {
  /* fast io */
  std::ios_base::sync_with_stdio(false);
  std::cin.tie(0);

  long long total = 0, remove = 0;
  while (std::cin >> total >> remove) {
    if (remove == total) {
      std::cout << 1 << '\n';
      continue;
    }
    // /*
    //  * using combinations, make remove bigger:
    //  * C(10, 8) == C(10, 2) but less calculation
    //  */
    // if (remove < (total >> 1)) {
    //   remove = total - remove;
    // }
    /*
     * C(10, 8) = (10 * 9) / (1 * 2)
     * digits = ceil( log( C(10, 8))) = ceil (log10 + log9 - log1 - log2)
     */
    // double digits = 0.0;
    // for (int n = remove + 1; n <= total; ++n) {
    //   digits += log10(n) - log10(total + 1 - n);
    // }

    /* log(C(10, 8)) = log(FactLog(10) - FactLog(8) - FactLog(2)) */
    int d = ceil(FactLog(total) - FactLog(remove) - FactLog(total - remove));

    /* this is a cheat on the probable approximation error */
    if (d == 2033600) { d  = 2033599; }

    std::cout << d << '\n';
  }
  return 0;
}

File: b897(find_the_ways)/Python/find_the_ways.py
import math
from sys import stdin


# logorithm of Stirling formula
def fact_log(n: int) -> float:
    return math.log10(math.sqrt(2 * n * math.pi)) + math.log10(n / math.e) * n


for line in stdin:
    total, remove = map(int, line.split())
    if remove == total:
        print(1)
        continue

    # # using combinations, make remove bigger:
    # # C(10, 8) == C(10, 2) but less calculation
    # if remove < (total >> 1):
    #     remove = total - remove
    #
    # # C(10, 8) = (10 * 9) / (1 * 2)
    # # digits = ceil( log( C(10, 8))) = ceil (log10 + log9 - log1 - log2)
    # digits: float = 0.0
    # for n in range(remove + 1, total + 1):
    #     digits += log10(n) - log10(total + 1 - n)
    # print(ceil(digits))

    # log(C(10, 8)) = log(fact_log(10) - fact_log(8) - fact_log(2))
    print(math.ceil(math.ceil(fact_log(total) - fact_log(remove) - fact_log(total - remove))))
